google-cloud-firestore/src/main/java/com/google/cloud/firestore/StreamableQuery.java

```diff
@@ -233,25 +262,61 @@ public void onCompleted() {
       return result;
     } catch (Exception error) {
       span.end(error);
+      metricsContext.recordEndToEndLatency(error);
       throw error;
     }
   }
 
+  class MonitoredStreamResponseObserver implements ApiStreamObserver<RunQueryResponse> {
+    private final ApiStreamObserver<RunQueryResponse> observer;
+    private final MetricsContext metricsContext;
+    private boolean receivedFirstResponse = false;
+
+    // Constructor to initialize with the delegate and MetricsContext
+    public MonitoredStreamResponseObserver(
+        ApiStreamObserver<RunQueryResponse> observer, MetricsContext metricsContext) {
+      this.observer = observer;
+      this.metricsContext = metricsContext;
+    }
+
+    @Override
+    public void onNext(RunQueryResponse value) {
+      if (!receivedFirstResponse) {
+        receivedFirstResponse = true;
+        metricsContext.recordFirstResponseLatency();
+      }
+      observer.onNext(value);
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      metricsContext.recordEndToEndLatency(t);
+      observer.onError(t);
+    }
+
+    @Override
+    public void onCompleted() {
+      metricsContext.recordEndToEndLatency();
+      observer.onCompleted();
+    }
+  }
+
   protected void internalStream(
-      final ApiStreamObserver<RunQueryResponse> runQueryResponseObserver,
+      final MonitoredStreamResponseObserver streamResponseObserver,
```

Comment:
there are 4 functions calling `internalStream`, if we handle the metrics recording outside `internalStream`, then it is repeated for 4 times. 
If we pass `metricsContext` into the `internalStream` directly, then there will be too many parameters.

The metrics call are basically in sync with the `ApiStreamObserver` calls (onResponse, onError, onCompleted), we can group them into one monitored stream observer class.  @ehsannas, what do you think if we integrate the tracing into the new class as well?

Reply:
that's a good idea 👍

Reply:
As discussed offline, we can do it in a separate PR.